From c594a385cebde6c1b65dd66d8e1727e4e2f5c75b Mon Sep 17 00:00:00 2001
From: PatrikLundell <j.patrik.r.lundell@gmail.com>
Date: Tue, 20 Aug 2024 20:34:30 +0200
Subject: [PATCH 083/125] Allow explosions outside of the reality bubble
 (#75567)

* Allow explosions outside of the reality bubble

* demanded unrelated change

* Block activation on explosion loaded maps

* Update src/explosion.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

* conflict...

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
---
 src/explosion.cpp | 225 ++++++++++++++++++++++++++++------------------
 src/explosion.h   |   8 +-
 src/map.cpp       |  22 ++---
 3 files changed, 157 insertions(+), 98 deletions(-)

diff --git a/src/explosion.cpp b/src/explosion.cpp
index da09dcf001..96d3aa43a9 100644
--- a/src/explosion.cpp
+++ b/src/explosion.cpp
@@ -22,6 +22,7 @@
 #include "cata_utility.h"
 #include "character.h"
 #include "color.h"
+#include "coordinates.h"
 #include "creature.h"
 #include "creature_tracker.h"
 #include "damage.h"
@@ -177,7 +178,7 @@ float gurney_spherical( const double charge, const double mass )
 }
 
 // (C1001) Compiler Internal Error on Visual Studio 2015 with Update 2
-static void do_blast( const Creature *source, const tripoint &p, const float power,
+static void do_blast( map *m, const Creature *source, const tripoint_bub_ms &p, const float power,
                       const float distance_factor, const bool fire )
 {
     const float tile_dist = 1.0f;
@@ -190,23 +191,22 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
     static constexpr std::array<int, 10> x_offset = { -1, 1,  0, 0,  1, -1, -1, 1, 0, 0 };
     static constexpr std::array<int, 10> y_offset = { 0, 0, -1, 1, -1,  1, -1, 1, 0, 0 };
     static constexpr std::array<int, 10> z_offset = { 0, 0,  0, 0,  0,  0,  0, 0, 1, -1 };
-    map &here = get_map();
     const size_t max_index = 10;
 
-    here.bash( p, fire ? power : ( 2 * power ), true, false, false );
+    m->bash( p, fire ? power : ( 2 * power ), true, false, false );
 
-    std::priority_queue< std::pair<float, tripoint>, std::vector< std::pair<float, tripoint> >, pair_greater_cmp_first >
+    std::priority_queue< std::pair<float, tripoint_bub_ms>, std::vector< std::pair<float, tripoint_bub_ms> >, pair_greater_cmp_first >
     open;
-    std::set<tripoint> closed;
-    std::set<tripoint> bashed{ p };
-    std::map<tripoint, float> dist_map;
+    std::set<tripoint_bub_ms> closed;
+    std::set<tripoint_bub_ms> bashed{ p };
+    std::map<tripoint_bub_ms, float> dist_map;
     open.emplace( 0.0f, p );
     dist_map[p] = 0.0f;
     // Find all points to blast
     while( !open.empty() ) {
         // Add some random factor to effective distance to make it look cooler
         const float distance = open.top().first * rng_float( 1.0f, 1.2f );
-        const tripoint pt = open.top().second;
+        const tripoint_bub_ms pt = open.top().second;
         open.pop();
 
         if( closed.count( pt ) != 0 ) {
@@ -220,7 +220,7 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
             continue;
         }
 
-        if( here.impassable( pt ) && pt != p ) {
+        if( m->impassable( pt ) && pt != p ) {
             // Don't propagate further
             continue;
         }
@@ -229,8 +229,8 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
         // Don't check up/down (for now) - this will make 2D/3D balancing easier
         int empty_neighbors = 0;
         for( size_t i = 0; i < 8; i++ ) {
-            tripoint dest( pt + tripoint( x_offset[i], y_offset[i], z_offset[i] ) );
-            if( closed.count( dest ) == 0 && here.valid_move( pt, dest, false, true ) ) {
+            tripoint_bub_ms dest( pt + tripoint_rel_ms( x_offset[i], y_offset[i], z_offset[i] ) );
+            if( closed.count( dest ) == 0 && m->valid_move( pt, dest, false, true ) ) {
                 empty_neighbors++;
             }
         }
@@ -238,8 +238,8 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
         empty_neighbors = std::max( 1, empty_neighbors );
         // Iterate over all neighbors. Bash all of them, propagate to some
         for( size_t i = 0; i < max_index; i++ ) {
-            tripoint dest( pt + tripoint( x_offset[i], y_offset[i], z_offset[i] ) );
-            if( closed.count( dest ) != 0 || !here.inbounds( dest ) ) {
+            tripoint_bub_ms dest( pt + tripoint_rel_ms( x_offset[i], y_offset[i], z_offset[i] ) );
+            if( closed.count( dest ) != 0 || !m->inbounds( dest ) ) {
                 continue;
             }
 
@@ -252,21 +252,21 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
                                          force / 2;
                 if( z_offset[i] == 0 ) {
                     // Horizontal - no floor bashing
-                    here.bash( dest, bash_force, true, false, false );
+                    m->bash( dest, bash_force, true, false, false );
                 } else if( z_offset[i] > 0 ) {
                     // Should actually bash through the floor first, but that's not really possible yet
-                    here.bash( dest, bash_force, true, false, true );
-                } else if( !here.valid_move( pt, dest, false, true ) ) {
+                    m->bash( dest, bash_force, true, false, true );
+                } else if( !m->valid_move( pt, dest, false, true ) ) {
                     // Only bash through floor if it doesn't exist
                     // Bash the current tile's floor, not the one's below
-                    here.bash( pt, bash_force, true, false, true );
+                    m->bash( pt, bash_force, true, false, true );
                 }
             }
 
             float next_dist = distance;
             next_dist += ( x_offset[i] == 0 || y_offset[i] == 0 ) ? tile_dist : diag_dist;
             if( z_offset[i] != 0 ) {
-                if( !here.valid_move( pt, dest, false, true ) ) {
+                if( !m->valid_move( pt, dest, false, true ) ) {
                     continue;
                 }
 
@@ -280,37 +280,46 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
         }
     }
 
-    // Draw the explosion
-    std::map<tripoint, nc_color> explosion_colors;
-    for( const tripoint &pt : closed ) {
-        if( here.impassable( pt ) ) {
-            continue;
-        }
+    // Draw the explosion, but only if the explosion center is within the reality bubble
+    map &bubble_map = get_map();
+    if( bubble_map.inbounds( m->getabs( p ) ) ) {
+        std::map<tripoint, nc_color> explosion_colors;
+        for( const tripoint_bub_ms &pt : closed ) {
+            const tripoint_bub_ms bubble_pos( bubble_map.bub_from_abs( m->getabs( pt ) ) );
 
-        const float force = power * std::pow( distance_factor, dist_map.at( pt ) );
-        nc_color col = c_red;
-        if( force < 10 ) {
-            col = c_white;
-        } else if( force < 30 ) {
-            col = c_yellow;
+            if( !bubble_map.inbounds( bubble_pos ) ) {
+                continue;
+            }
+            if( m->impassable( pt ) ) {
+                continue;
+            }
+
+            const float force = power * std::pow( distance_factor, dist_map.at( pt ) );
+            nc_color col = c_red;
+            if( force < 10 ) {
+                col = c_white;
+            } else if( force < 30 ) {
+                col = c_yellow;
+            }
+
+            explosion_colors[bubble_pos.raw()] = col;
         }
 
-        explosion_colors[pt] = col;
+        draw_custom_explosion( get_player_character().pos(), explosion_colors );
     }
 
-    draw_custom_explosion( get_player_character().pos(), explosion_colors );
-
     creature_tracker &creatures = get_creature_tracker();
-    Creature *mutable_source = source == nullptr ? nullptr : creatures.creature_at( source->pos() );
-    for( const tripoint &pt : closed ) {
+    // Must use the reality bubble pos, because that's what the creature tracker works with.
+    Creature *mutable_source = source == nullptr ? nullptr : creatures.creature_at( source->pos_bub() );
+    for( const tripoint_bub_ms &pt : closed ) {
         const float force = power * std::pow( distance_factor, dist_map.at( pt ) );
         if( force < 1.0f ) {
             // Too weak to matter
             continue;
         }
 
-        if( here.has_items( pt ) ) {
-            here.smash_items( pt, force, _( "force of the explosion" ) );
+        if( m->has_items( pt ) ) {
+            m->smash_items( pt, force, _( "force of the explosion" ) );
         }
 
         if( fire ) {
@@ -318,16 +327,18 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
             if( force > 10.0f || x_in_y( force, 10.0f ) ) {
                 intensity++;
             }
-            here.add_field( pt, fd_fire, intensity );
+            m->add_field( pt, fd_fire, intensity );
         }
 
-        if( const optional_vpart_position vp = here.veh_at( pt ) ) {
+        if( const optional_vpart_position vp = m->veh_at( pt ) ) {
             // TODO: Make this weird unit used by vehicle::damage more sensible
-            vp->vehicle().damage( here, vp->part_index(), force,
+            vp->vehicle().damage( m[0], vp->part_index(), force,
                                   fire ? damage_heat : damage_bash, false );
         }
 
-        Creature *critter = creatures.creature_at( pt, true );
+        // Translate to reality bubble coordinates to work with the creature tracker.
+        const tripoint_bub_ms bubble_pos( bubble_map.bub_from_abs( m->getabs( pt ) ) );
+        Creature *critter = creatures.creature_at( bubble_pos, true );
         if( critter == nullptr ) {
             continue;
         }
@@ -348,8 +359,10 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
         }
 
         // Print messages for all NPCs
-        pl->add_msg_player_or_npc( m_bad, _( "You're caught in the explosion!" ),
-                                   _( "<npcname> is caught in the explosion!" ) );
+        if( bubble_map.inbounds( bubble_pos ) ) {
+            pl->add_msg_player_or_npc( m_bad, _( "You're caught in the explosion!" ),
+                                       _( "<npcname> is caught in the explosion!" ) );
+        }
 
         struct blastable_part {
             bodypart_id bp;
@@ -386,8 +399,9 @@ static void do_blast( const Creature *source, const tripoint &p, const float pow
     }
 }
 
-static std::vector<tripoint> shrapnel( const Creature *source, const tripoint &src, int power,
-                                       int casing_mass, float per_fragment_mass, int range = -1 )
+static std::vector<tripoint_bub_ms> shrapnel( map *m, const Creature *source,
+        const tripoint_bub_ms &src, int power,
+        int casing_mass, float per_fragment_mass, int range = -1 )
 {
     // The gurney equation wants the total mass of the casing.
     const float fragment_velocity = gurney_spherical( power, casing_mass );
@@ -396,7 +410,7 @@ static std::vector<tripoint> shrapnel( const Creature *source, const tripoint &s
     int fragment_count = casing_mass / fragment_mass;
 
     // Contains all tiles reached by fragments.
-    std::vector<tripoint> distrib;
+    std::vector<tripoint_bub_ms> distrib;
 
     projectile proj;
     proj.speed = fragment_velocity;
@@ -412,38 +426,41 @@ static std::vector<tripoint> shrapnel( const Creature *source, const tripoint &s
     cata::mdarray<fragment_cloud, point_bub_ms> &obstacle_cache = caches->obstacle_cache;
     cata::mdarray<fragment_cloud, point_bub_ms> &visited_cache = caches->visited_cache;
 
-    map &here = get_map();
     // TODO: Calculate range based on max effective range for projectiles.
     // Basically bisect between 0 and map diameter using shrapnel_calc().
     // Need to update shadowcasting to support limiting range without adjusting initial distance.
-    const tripoint_range<tripoint> area = here.points_on_zlevel( src.z );
+    const tripoint_range<tripoint_bub_ms> area = m->bub_points_on_zlevel( src.z() );
 
-    here.build_obstacle_cache( area.min(), area.max() + tripoint_south_east, obstacle_cache );
+    m->build_obstacle_cache( area.min().raw(), area.max().raw() + tripoint_south_east, obstacle_cache );
 
     // Shadowcasting normally ignores the origin square,
     // so apply it manually to catch monsters standing on the explosive.
     // This "blocks" some fragments, but does not apply deceleration.
-    fragment_cloud initial_cloud = accumulate_fragment_cloud( obstacle_cache[src.x][src.y],
+    fragment_cloud initial_cloud = accumulate_fragment_cloud( obstacle_cache[src.x()][src.y()],
     { fragment_velocity, static_cast<float>( fragment_count ) }, 1 );
-    visited_cache[src.x][src.y] = initial_cloud;
-    visited_cache[src.x][src.y].density = static_cast<float>( fragment_count / 2.0 );
+    visited_cache[src.x()][src.y()] = initial_cloud;
+    visited_cache[src.x()][src.y()].density = static_cast<float>( fragment_count / 2.0 );
 
     castLightAll<fragment_cloud, fragment_cloud, shrapnel_calc, shrapnel_check,
                  update_fragment_cloud, accumulate_fragment_cloud>
-                 ( visited_cache, obstacle_cache, point_bub_ms( src.xy() ), 0, initial_cloud );
+                 ( visited_cache, obstacle_cache, src.xy(), 0, initial_cloud );
 
     creature_tracker &creatures = get_creature_tracker();
-    Creature *mutable_source = source == nullptr ? nullptr : creatures.creature_at( source->pos() );
+    // Creature tracker works on reality bubble coordinates, so feeding it with those coordinates from the critter is correct.
+    Creature *mutable_source = source == nullptr ? nullptr : creatures.creature_at( source->pos_bub() );
+    map &bubble_map = get_map();
     // Now visited_caches are populated with density and velocity of fragments.
-    for( const tripoint &target : area ) {
-        fragment_cloud &cloud = visited_cache[target.x][target.y];
+    for( const tripoint_bub_ms &target : area ) {
+        fragment_cloud &cloud = visited_cache[target.x()][target.y()];
         if( cloud.density <= MIN_FRAGMENT_DENSITY ||
             cloud.velocity <= MIN_EFFECTIVE_VELOCITY ) {
             continue;
         }
         distrib.emplace_back( target );
         int damage = ballistic_damage( cloud.velocity, fragment_mass );
-        Creature *critter = creatures.creature_at( target );
+        // Translate to reality bubble coordinates to work with the creature tracker.
+        const tripoint_bub_ms bubble_pos( bubble_map.bub_from_abs( m->getabs( target ) ) );
+        Creature *critter = creatures.creature_at( bubble_pos );
         if( damage > 0 && critter && !critter->is_dead_state() ) {
             std::poisson_distribution<> d( cloud.density );
             int hits = d( rng_get_engine() );
@@ -476,14 +493,17 @@ static std::vector<tripoint> shrapnel( const Creature *source, const tripoint &s
                 }
             }
             int total_hits = damaging_hits + non_damaging_hits;
-            multi_projectile_hit_message( critter, total_hits, damage_taken, n_gettext( "bomb fragment",
-                                          "bomb fragments", total_hits ) );
+            if( bubble_map.inbounds(
+                    bubble_pos ) ) { // Only report on critters in the reality bubble. Should probably be only for visible critters...
+                multi_projectile_hit_message( critter, total_hits, damage_taken, n_gettext( "bomb fragment",
+                                              "bomb fragments", total_hits ) );
+            }
         }
-        if( here.impassable( target ) ) {
-            if( optional_vpart_position vp = here.veh_at( target ) ) {
-                vp->vehicle().damage( here, vp->part_index(), damage / 10 );
+        if( m->impassable( target ) ) {
+            if( optional_vpart_position vp = m->veh_at( target ) ) {
+                vp->vehicle().damage( m[0], vp->part_index(), damage / 10 );
             } else {
-                here.bash( target, damage / 100, true );
+                m->bash( target, damage / 100, true );
             }
         }
     }
@@ -503,23 +523,41 @@ void explosion( const Creature *source, const tripoint &p, float power, float fa
     explosion( source, p, data );
 }
 
+// Blocks activation of maps loaded by process_explosions. Activation would trigger a recursive
+// activation of whatever caused the explosion triggering the map loading, leading to a recursive
+// death spiral. Activation could also trigger further explosions leading to a cascade effect which
+// is also undesirable. Such explosions should not be triggered recursively, but by a direct action.
+static bool process_explosions_in_progress = false;
+
+bool explosion_processing_active()
+{
+    return process_explosions_in_progress;
+}
+
 void explosion( const Creature *source, const tripoint &p, const explosion_data &ex )
 {
     _explosions.emplace_back( source, get_map().getglobal( p ), ex );
 }
 
-void _make_explosion( const Creature *source, const tripoint &p, const explosion_data &ex )
+void _make_explosion( map *m, const Creature *source, const tripoint_bub_ms &p,
+                      const explosion_data &ex )
 {
-    int noise = ex.power * ( ex.fire ? 2 : 10 );
-    noise = ( noise > ex.max_noise ) ? ex.max_noise : noise;
-    if( noise >= 30 ) {
-        sounds::sound( p, noise, sounds::sound_t::combat, _( "a huge explosion!" ), false, "explosion",
-                       "huge" );
-    } else if( noise >= 4 ) {
-        sounds::sound( p, noise, sounds::sound_t::combat, _( "an explosion!" ), false, "explosion",
-                       "default" );
-    } else if( noise > 0 ) {
-        sounds::sound( p, 3, sounds::sound_t::combat, _( "a loud pop!" ), false, "explosion", "small" );
+    if( get_map().inbounds( m->getabs( p ) ) ) {
+        tripoint_bub_ms bubble_pos = get_map().bub_from_abs( m->getabs( p ) );
+        int noise = ex.power * ( ex.fire ? 2 : 10 );
+        noise = ( noise > ex.max_noise ) ? ex.max_noise : noise;
+
+        if( noise >= 30 ) {
+            sounds::sound( bubble_pos, noise, sounds::sound_t::combat, _( "a huge explosion!" ), false,
+                           "explosion",
+                           "huge" );
+        } else if( noise >= 4 ) {
+            sounds::sound( bubble_pos, noise, sounds::sound_t::combat, _( "an explosion!" ), false, "explosion",
+                           "default" );
+        } else if( noise > 0 ) {
+            sounds::sound( bubble_pos, 3, sounds::sound_t::combat, _( "a loud pop!" ), false, "explosion",
+                           "small" );
+        }
     }
 
     if( ex.distance_factor >= 1.0f ) {
@@ -527,21 +565,21 @@ void _make_explosion( const Creature *source, const tripoint &p, const explosion
     } else if( ex.distance_factor > 0.0f && ex.power > 0.0f ) {
         // Power rescaled to mean grams of TNT equivalent, this scales it roughly back to where
         // it was before until we re-do blasting power to be based on TNT-equivalent directly.
-        do_blast( source, p, ex.power / 15.0, ex.distance_factor, ex.fire );
+        do_blast( m, source, p, ex.power / 15.0, ex.distance_factor, ex.fire );
     }
 
-    map &here = get_map();
     const shrapnel_data &shr = ex.shrapnel;
     if( shr.casing_mass > 0 ) {
-        auto shrapnel_locations = shrapnel( source, p, ex.power, shr.casing_mass, shr.fragment_mass );
+        std::vector<tripoint_bub_ms> shrapnel_locations = shrapnel( m, source, p, ex.power, shr.casing_mass,
+                shr.fragment_mass );
 
         // If explosion drops shrapnel...
         if( shr.recovery > 0 && !shr.drop.is_null() ) {
 
             // Extract only passable tiles affected by shrapnel
-            std::vector<tripoint> tiles;
-            for( const tripoint &e : shrapnel_locations ) {
-                if( here.passable( e ) ) {
+            std::vector<tripoint_bub_ms> tiles;
+            for( const tripoint_bub_ms &e : shrapnel_locations ) {
+                if( m->passable( e ) ) {
                     tiles.push_back( e );
                 }
             }
@@ -552,8 +590,8 @@ void _make_explosion( const Creature *source, const tripoint &p, const explosion
             std::shuffle( tiles.begin(), tiles.end(), rng_get_engine() );
             tiles.resize( std::min( static_cast<int>( tiles.size() ), qty ) );
 
-            for( const tripoint &e : tiles ) {
-                here.add_item_or_charges( e, item( shr.drop, calendar::turn, item::solitary_tag{} ) );
+            for( const tripoint_bub_ms &e : tiles ) {
+                m->add_item_or_charges( e, item( shr.drop, calendar::turn, item::solitary_tag{} ) );
             }
         }
     }
@@ -927,12 +965,25 @@ void process_explosions()
     _explosions.clear();
 
     for( const queued_explosion &ex : explosions_copy ) {
-        const tripoint p = get_map().getlocal( ex.pos );
-        if( p.x < 0 || p.x >= MAPSIZE_X || p.y < 0 || p.y >= MAPSIZE_Y ) {
-            debugmsg( "Explosion origin (%d, %d, %d) is out-of-bounds", p.x, p.y, p.z );
-            continue;
+        const int safe_range = ex.data.safe_range();
+        map  *bubble_map = &get_map();
+        const tripoint_bub_ms bubble_pos( bubble_map->bub_from_abs( ex.pos ) );
+
+        if( bubble_pos.x() - safe_range < 0 || bubble_pos.x() + safe_range > MAPSIZE_X ||
+            bubble_pos.y() - safe_range < 0 || bubble_pos.y() + safe_range > MAPSIZE_Y ) {
+            map m;
+            const tripoint_abs_sm origo( project_to<coords::sm>( ex.pos ) - point_rel_sm{ HALF_MAPSIZE, HALF_MAPSIZE} );
+            // Create a map centered around the explosion point to allow an explosion with a radius of up to 5 submaps
+            // to be created without being cut off by the map's boundary. That also means there is no need for the map
+            // to actually overlap the reality bubble, so a large explosion can be detonated without blowing up the PC
+            // or have a vehicle run into a crater suddenly appearing just in front of it.
+            process_explosions_in_progress = true;
+            m.load( origo, false, false );
+            process_explosions_in_progress = false;
+            _make_explosion( &m, ex.source, m.bub_from_abs( ex.pos ), ex.data );
+        } else {
+            _make_explosion( bubble_map, ex.source, bubble_map->bub_from_abs( ex.pos ), ex.data );
         }
-        _make_explosion( ex.source, p, ex.data );
     }
 }
 
diff --git a/src/explosion.h b/src/explosion.h
index 5b0bc512c0..fd1dda77ba 100644
--- a/src/explosion.h
+++ b/src/explosion.h
@@ -8,6 +8,7 @@
 #include <vector>
 
 #include "coordinates.h"
+#include "map.h"
 #include "type_id.h"
 
 class Creature;
@@ -78,8 +79,13 @@ void explosion(
     bool fire = false, int casing_mass = 0, float frag_mass = 0.05
 );
 
+// Explosion processing is loading a map on which to execute the explosion. Processing that
+// would potentially set off additional explosions should not be performed. They should wait
+// until triggered normally.
+bool explosion_processing_active();
 void explosion( const Creature *source, const tripoint &p, const explosion_data &ex );
-void _make_explosion( const Creature *source, const tripoint &p, const explosion_data &ex );
+void _make_explosion( map *m, const Creature *source, const tripoint_bub_ms &p,
+                      const explosion_data &ex );
 
 /** Triggers a flashbang explosion at p. */
 void flashbang( const tripoint &p, bool player_immune = false );
diff --git a/src/map.cpp b/src/map.cpp
index dc8c2cbea7..566af69b02 100644
--- a/src/map.cpp
+++ b/src/map.cpp
@@ -8377,16 +8377,18 @@ void map::load( const tripoint_abs_sm &w, const bool update_vehicle,
     }
     rebuild_vehicle_level_caches();
 
-    // actualize after loading all submaps to prevent errors
-    // with entities at the edges
-    for( int gridx = 0; gridx < my_MAPSIZE; gridx++ ) {
-        for( int gridy = 0; gridy < my_MAPSIZE; gridy++ ) {
-            const int zmin = zlevels ? -OVERMAP_DEPTH : abs_sub.z();
-            const int zmax = zlevels ? OVERMAP_HEIGHT : abs_sub.z();
-            for( int gridz = zmin; gridz <= zmax; gridz++ ) {
-                actualize( {gridx, gridy, gridz            } );
-                if( pump_events ) {
-                    inp_mngr.pump_events();
+    if( !explosion_handler::explosion_processing_active() ) {
+        // actualize after loading all submaps to prevent errors
+        // with entities at the edges
+        for( int gridx = 0; gridx < my_MAPSIZE; gridx++ ) {
+            for( int gridy = 0; gridy < my_MAPSIZE; gridy++ ) {
+                const int zmin = zlevels ? -OVERMAP_DEPTH : abs_sub.z();
+                const int zmax = zlevels ? OVERMAP_HEIGHT : abs_sub.z();
+                for( int gridz = zmin; gridz <= zmax; gridz++ ) {
+                    actualize( { gridx, gridy, gridz } );
+                    if( pump_events ) {
+                        inp_mngr.pump_events();
+                    }
                 }
             }
         }
-- 
2.45.2

